### PR TITLE
Implement foldIterate

### PR DESCRIPTION
**Pull request by adithyaov**
Benchmarks,
```
benchmarked Data.Fold/o-1-space/nesting/foldIterateM (take 1)
time                 115.4 μs   (114.6 μs .. 116.4 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 116.8 μs   (116.2 μs .. 117.3 μs)
std dev              2.446 μs   (2.150 μs .. 2.918 μs)
variance introduced by outliers: 12% (moderately inflated)

benchmarked Data.Fold/o-1-space/nesting/foldIterateM (take all)
time                 817.1 μs   (812.2 μs .. 821.7 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 816.6 μs   (813.0 μs .. 820.2 μs)
std dev              12.36 μs   (10.22 μs .. 15.09 μs)
```
These benchmarks look incorrect,
1. I expected `foldIterateM (take 1)` to be worse compared to `foldIterateM (take all)`
2. In contrast `parseIterate` is much worse. I expected them to be almost the same.
```
benchmarked Data.Parser/o-1-space/parseIterate (take 1)
time                 3.046 ms   (3.025 ms .. 3.059 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 3.035 ms   (3.026 ms .. 3.052 ms)
std dev              31.09 μs   (15.04 μs .. 48.53 μs)

benchmarked Data.Parser/o-1-space/parseIterate (take all)
time                 3.239 ms   (3.224 ms .. 3.252 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 3.242 ms   (3.232 ms .. 3.253 ms)
std dev              26.36 μs   (19.39 μs .. 35.92 μs)
```
Looks like my hunch about `take 1` being worse compared to `take all` is incorrect. Why though?

### CODE REVIEW

**Review by harendra-kumar — APPROVED** [review]

File: benchmark/Streamly/Benchmark/Data/Fold.hs [hunk 1]
```diff
@@ -176,13 +176,34 @@ classifyWith f = S.fold (FL.classifyWith f FL.sum)
...
+-------------------------------------------------------------------------------
+
+{-# INLINE foldIterateM #-}
+foldIterateM :: Monad m => Int -> SerialT m Int -> m ()
+foldIterateM n =
+    S.drain
+        . S.map getSum
+        . IP.foldIterateM (return . FL.take n . FL.sconcat) (Sum 0)
+        . S.map Sum
+
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Fold.hs:190]: Move this to stream benchmarks.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Comment by adithyaov** [comment]
```
benchmarked Prelude.Serial/o-1-space/grouping/foldIterateM
time                 466.0 μs   (462.2 μs .. 469.1 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 461.9 μs   (460.0 μs .. 464.3 μs)
std dev              8.128 μs   (7.059 μs .. 9.759 μs)
```

File: benchmark/Streamly/Benchmark/Data/Fold.hs (hunk 1 above)
Comment (adithyaov) [reply benchmark/Streamly/Benchmark/Data/Fold.hs]: A benchmark already existed. Uncommented it.

File: test/Streamly/Test/Data/Fold.hs [hunk 2]
```diff
@@ -461,3 +480,4 @@ main = hspec $
         prop "distribute" Main.distribute
         prop "partition" Main.partition
         prop "unzip" Main.unzip
+        prop "foldIterate" Main.foldIterate
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Fold.hs:483]: The test should go in prelude tests, as the operation is in Prelude and not in Data.Fold.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: test/Streamly/Test/Data/Fold.hs (hunk 2 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Fold.hs:483]: I've mimicked the parser layout. I remember seeing `parseIterate` and `parseMany` in `bench:Data.Parser`

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**